### PR TITLE
ENH: Add DiscreteGuideTable wrapper

### PR DESCRIPTION
**Pull request by Kai-Striega**
#### Reference issue 
gh-14600

#### What does this implement/fix? 
This is a very minimal start of to the DiscreteGuideTable wrapper. It is included in `all_methods` and the associated tests pass, however additional tests, documentation and type hinting must still be added.

#### Additional information
I'm having some problems having the wrapper recognize the probability vector memory view, this can be reproduced as follows: 
```python
import numpy as np
from scipy.stats import DiscreteGuideTable

pv = [0.1, 0.3, 0.6]
urng = np.random.default_rng()
rng = DiscreteGuideTable(pv, random_state=urng)
Traceback (most recent call last): File "", line 1, in File "unuran_wrapper.pyx", line 1632, in scipy.stats._unuran.unuran_wrapper.DiscreteGuideTable.__cinit__ AttributeError: 'scipy.stats._unuran.unuran_wrapper.DiscreteGuideTable' object has no attribute 'pv_view'
```
It's also interesting to note that the tests still pass despite this.

### CODE REVIEW

**Comment by Kai-Striega** [comment]
@tirthasheshpatel 

I'm having some problems with importing ``DiscreteGuideTable`` after your changes, could you take a quick look if it's something obvious?

```python
from scipy.stats import DiscreteGuideTable
Traceback (most recent call last):
  File "/home/kai/anaconda3/envs/scipy-dev/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3444, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-1cc9022332b4>", line 1, in <module>
    from scipy.stats import DiscreteGuideTable
  File "/home/kai/.local/share/JetBrains/Toolbox/apps/PyCharm-P/ch-0/212.5284.44/plugins/python/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/kai/PycharmProjects/scipy/scipy/stats/__init__.py", line 463, in <module>
    from ._unuran import *  # noqa
```

I haven't made any relevant changes so I suspect it's introduced by the code you merged, although I'm not 100% sure.

**Review by tirthasheshpatel — COMMENTED** [review]
Some comments on typos, style, and tests. Wrapper code, docs, and tests look very good! You can copy-paste the documentation in the tutorial and write a small benchmark for the method to finish this up. Given the similarity to `DiscreteAliasUrn`, I think we should be good to go if tuts and benchmarks look good. Can you take a look whenever you are free @chrisb83?

**Review by chrisb83 — COMMENTED** [review]
I just noted a few minor inconsistencies/typos.

The only point we should decide is whether we need the parameter `variant`. I think we could drop it and go with the default recommendation of the UNU.RAN authors (i.e., the current case of None ). We do not have much information about what the difference is between 1 and 2, so maybe better to simply hide this from the user?

**Review by chrisb83 — COMMENTED** [review]
I resolved all open points, just two minor comments remain to be addressed

**Review by chrisb83 — COMMENTED** [review]
I found two remaining occurrences of `variant`. Once this is resolved, I think I will approve the PR.

**Review by chrisb83 — APPROVED** [review]
OK, the doc build works now. Thanks Kai? No further comments from my side. @tirthasheshpatel if you are fine with the changes, feel free to merge the PR.

**Review by tirthasheshpatel — COMMENTED** [review]
Thanks for the changes so far, @Kai-Striega! One last thing:

DGT is an inversion method and UNU.RAN provides the function `unur_quantile` that can compute the PPF using generators of type DGT. So, we can add a `ppf` method that can provide exact PPF (because this is a discrete distribution) of the given distribution. This is especially a desired method if there is no closed-form solution to the inverse of the CDF or it is too expensive to compute it. Can you please add a `ppf` method and a test for it? You can look at `NumericalInversePolynomial` for an example of how to implement the `ppf` method.

**Comment by Kai-Striega** [comment]
Here is the ``_ppf`` method for ``NumericalInversePolynomial``:

```python
@cython.boundscheck(False)
@cython.wraparound(False)
cdef inline void _ppf(self, const double *u, double *out, size_t N):
    cdef:
        size_t i
    for i in range(N):
        out[i] = unur_pinv_eval_approxinvcdf(self.rng, u[i])
```

I'm assuming that the call to ``unur_pinv_eval_approxinvcdf`` must be replaced with one specific to DGT, but from where to I find the available method names? Is there an index of all available names?

**Comment by chrisb83** [comment]
It is described here: https://statmath.wu.ac.at/unuran/doc/unuran.html#funct_003aunur_005fquantile

**Comment by Kai-Striega** [comment]
@tirthasheshpatel I've got it running but there's a bug (I think) in SciPy's existing implementation.

```python
>>> from scipy import stats
>>> stats.binom.ppf(0, 4, 0.1)
-1.0
```

shouldn't this return 0? In contrast:

```python
>>> n, p = 4, 0.1
>>> dist = Binomial(n, p)
>>> rng = DiscreteGuideTable(dist, random_state=42)
>>> rng.ppf(0)
0.
```

There's also some issues with the example generated from ``np.linspace`` not being exactly the same for various numbers.

**Comment by tirthasheshpatel** [comment]
> shouldn't this return 0?

AFAIK, there are two ways to handle the edge case `u = 0`:

1. Fill in the left boundary of the domain when `u = 0`.
2. Fill in `left_domain - 1` when `u = 0`.

SciPy prefers the second option while UNU.RAN prefers the first. I have tried to be consistent with SciPy implementation in `NumericalInverseHermite` and `NumericalInversePolynomial` so we should fallback to SciPy's behavior for the case `u = 0` for discrete distributions.

> There's also some issues with the example generated from `np.linspace` not being exactly the same for various numbers.

Can you share an example? I assume it might be because the generator's PPF is just an approximation.

**Comment by Kai-Striega** [comment]
> Can you share an example? I assume it might be because the generator's PPF is just an approximation.
```python
>>> n, p = 4, 0.1
>>> dist = Binomial(n, p)  # as defined in tests
>>> rng = DiscreteGuideTable(dist, random_state=42)
>>> rng.ppf(np.linspace(0, 1, num=10))
np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 4.])
>>> stats.binom.ppf(np.linspace(0, 1, num=10), n, p)
np.array([-1.,  0.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  4.])
```

This could just be because it's an approximation, but it still makes it hard to test.

**Comment by tirthasheshpatel** [comment]
> This could just be because it's an approximation, but it still makes it hard to test.

Actually, there's a bug in `Binomial` class defined in tests. The binomial coefficient is missing in the PMF method (which I, at the time of coding it, thought is a normalizing constant). Right now it is used at only one place in `DiscreteAliasUrn`'s test suite. It would be best to remove it and use the equivalent `binom` class instead.

**Comment by Kai-Striega** [comment]
> It would be best to remove it and use the equivalent binom class instead.

Do you want me to do that for this PR? I'm happy to use the stats version in my test but I'm worried that there's already a lot happening in the current changes and don't really want to add changes from DiscreteAlaisUrn too.

**Comment by tirthasheshpatel** [comment]
> Actually this should be a pretty small change so I'll take a look at adding it

Yeah, that's also why I think it's OK to change that here.

**Comment by Kai-Striega** [comment]
Ok. Give me ~30 minutes and this should be good to go.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2224]: SciPy fills in `left_boundary - 1` for `u = 0`. So, if we have `dist = binom(4, 0.1, loc=10)`, it will give `9.0` for `dist.ppf(0)`. So, we should save the `domain` in a `self.domain` attribute in the `__cinit__` method and use `self.domain[0]-1` to fill in the `u = 0` case.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2080]: Can you also add an example of how to use the `ppf` method?

File: scipy/stats/_unuran/unuran_wrapper.pyi
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyi:136]: we need to add type hints for the `ppf` method.

File: scipy/stats/tests/test_sampling.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_sampling.py:916]: As the PPF is exact, we should use `assert_equal` for a more robust test that checks if the generator's PPF matches exactly with the true PPF.

**Review by tirthasheshpatel — COMMENTED** [review]
Thanks for adding the `ppf` method. The test looks good. Some comments on adding examples and type hints.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1971]: minor point: use single backticks `guide_factor`

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2190]: remove Approximated since DGT returns the exact value

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2195]: typo

**Review by chrisb83 — COMMENTED** [review]
I observed a few more minor things but we are very close :)

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (Kai-Striega) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:2224]: So that would be: ``np.place(out, u == 0, self.domain[0]-1)``?

Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:2224]: > So that would be: `np.place(out, u == 0, self.domain[0]-1)`, but also storing the domain in self.domain

Yes, exactly.

**Comment by chrisb83** [comment]
@Kai-Striega Hi Kai, do you think you have time to work on the PR this weekend? We are very close to a final version and it would be great to get this merged soon.

**Comment by Kai-Striega** [comment]
I'll get it done tomorrow. I've been sick this week and I've had to prioritise catching up on my work

File: scipy/stats/tests/test_sampling.py
Comment (Kai-Striega) [reply scipy/stats/tests/test_sampling.py]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyi
Comment (Kai-Striega) [reply scipy/stats/_unuran/unuran_wrapper.pyi:136]: Done. But I'm not 100% sure if it's done correctly, would you please double check?

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (Kai-Striega) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:2079]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (Kai-Striega) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (Kai-Striega) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Fixed here and in ``NumericalInversePolynomial``

File: scipy/stats/_unuran/unuran_wrapper.pyi
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyi:144]: Needs an overload for scalar inputs.

```suggestion
    @overload
    def ppf(self, u: ArrayLike0D) -> float: ...
    @overload
    def ppf(self, u: npt.ArrayLike) -> np.ndarray: ...
```

**Comment by Kai-Striega** [comment]
@tirthasheshpatel could you take a look at why there are merge issues. I don't *think* that it's anything I did but I'm not sure what's causing them

**Comment by tirthasheshpatel** [comment]
Thanks for the changes, @Kai-Striega! I think the last thing left to do is return `domain[0]-1` when `ppf` is evaluated at `u = 0`. Also, can you add the `ppf` example to the tutorial? Don't worry about the conflicts. We can resolve them before merging.

**Comment by Kai-Striega** [comment]
> I think the last thing left to do is ...

There's also one more instance of guide_factor that requires back-ticks around it

> return domain[0]-1 when ppf is evaluated at u = 0

I'm having some weird issue with this.  I've added the line ``self.domain = domain`` on line 2101 but, even without any changes to ``ppf`` I'm getting the error ``AttributeError: 'scipy.stats._unuran.unuran_wrapper.DiscreteGuideTable' object has no attribute 'domain'``. I'm going to try an rebuild it from scratch to see if that helps otherwise it may take some time to figure out why I can't add that attribute.

> Also, can you add the ppf example to the tutorial?

This should already be done starting on line 102 of `doc/source/tutorial/stats/sampling_dgt.rst`

> Don't worry about the conflicts. We can resolve them before merging.

ok. I just don't normally like not having the tests run

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2090]: @Kai-Striega You will have to declare the attribute `cdef object domain` before assigning to it in the `__cinit__` method:

```suggestion
    cdef double[::1] pv_view
    cdef object domain
```

**Comment by tirthasheshpatel** [comment]
> I just don't normally like not having the tests run

Oh, yes! Let me know once you are done with the `domain` change. I will help resolve conflicts afterwards.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (Kai-Striega) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:2090]: Thank you! I haven't used cython much so I'm coming across some of these "silly" mistakes

**Comment by Kai-Striega** [comment]
@tirthasheshpatel it should be ready now ;)

File: scipy/stats/_unuran/unuran_wrapper.pyi
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyi:178]:
```suggestion
    def ppf(self, u: ArrayLike0D) -> float: ...  # type: ignore[misc]
```

File: scipy/stats/_unuran/unuran_wrapper.pyi
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyi:180]:
```suggestion
    @overload
    def ppf(self, u: npt.ArrayLike) -> np.ndarray: ...  # type: ignore[misc]
```

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2849]: If a probability vector is given without a domain, we assume the domain starts from 0. In this case, the attribute `self.domain` is `None` and hence not subscribable. So, we can fill in `-1` directly in this case:

```suggestion
        # UNU.RAN sets boundary at u = 0 to domain[0]
        # SciPy fills it with domain[0] - 1. Prefer SciPy behaviour
        if domain is not None:
            np.place(out, u == 0, self.domain[0] - 1)
        else:
            # domain starts at 0. So, fill in -1.
            np.place(out, u == 0, -1)
```

**Review by tirthasheshpatel — COMMENTED** [review]
This is very close! Last batch of comments to resolve the failing tests.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2849]: Opps. This should be `self.domain`:

```suggestion
        if self.domain is not None:
```

**Comment by Kai-Striega** [comment]
@tirthasheshpatel I've fixed most of the CI failures, there is one more on MyPy that I can't fix - short of removing the line in its entirety. Could you please take a look and check if there is any better way to get it to pass?

File: scipy/stats/_unuran/unuran_wrapper.pyi
Comment (tirthasheshpatel) [inline scipy/stats/_unuran/unuran_wrapper.pyi:180]:
```suggestion
    def ppf(self, u: npt.ArrayLike) -> np.ndarray: ...
```

File: doc/source/tutorial/stats/sampling_dgt.rst
Comment (tirthasheshpatel) [inline doc/source/tutorial/stats/sampling_dgt.rst:104]: This method doesn't address both shortcomings i.e. only provides `ppf` and not `cdf`. I think we should remove the sentence about CDF.

**Review by tirthasheshpatel — APPROVED** [review]
Everything looks good to me now! Tests are comprehensive; docs and tutorials are informative. Only one comment on the tutorial above. @chrisb83 Let me know if we still have your approval after the new changes. If you are happy with it, feel free to merge!

File: doc/source/tutorial/stats/sampling_dgt.rst
Comment (Kai-Striega) [reply doc/source/tutorial/stats/sampling_dgt.rst:104]:
```suggestion
available. The user only has to provide the
```

File: doc/source/tutorial/stats/sampling_dgt.rst
Comment (Kai-Striega) [inline doc/source/tutorial/stats/sampling_dgt.rst:106]:
```suggestion
available. The user only has to provide the probability vector and the 
PPF (inverse CDF) may be generated using ``ppf`` method. This 
method calculates the PPF exactly.
```

Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_dgt.rst:106]: How about we rephrase it this way:

```suggestion
available. The user only has to provide the probability vector and the 
PPF (inverse CDF) can be evaluated using ``ppf`` method. This 
method calculates the (exact) PPF of the given distribution.
```

Comment (Kai-Striega) [reply doc/source/tutorial/stats/sampling_dgt.rst:106]: I think that sounds better. I was simply deleting the incorrect pieces.

**Comment by Kai-Striega** [comment]
@tirthasheshpatel @chrisb83 I'd like to thank both of you for all the help - and patience - on this PR. It is appreciated!

File: doc/source/tutorial/stats/sampling_dgt.rst
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_dgt.rst]: I have committed this. I think this PR is ready now!

**Comment by tirthasheshpatel** [comment]
Thanks for your help too @Kai-Striega! This was a big one :)

**Comment by chrisb83** [comment]
Thanks Kai and Tirth! Nice to have this method in SciPy now

**Comment by tirthasheshpatel** [comment]
Thanks again @Kai-Striega! And thanks @chrisb83 for the reviews!

**Pull request merged**
